workers/unity/Assets/Gdk/Core/Components/Option.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+
+namespace Improbable.Gdk.Core
+{
+    public struct Option<T> : IEquatable<Option<T>>
+    {
+        public static readonly Option<T> Empty = new Option<T>();
+
+        public BlittableBool HasValue { get; }
+
+        private readonly T value;
+
+        public T Value
+        {
+            get
+            {
+                if (!HasValue)
+                {
+                    throw new CalledValueOnEmptyOptionException("Called Value on empty Option.");
+                }
+
+                return value;
+            }
+        }
+
+        public Option(T value)
+        {
+            HasValue = true;
+            this.value = value;
+        }
+
+        public bool TryGetValue(out T value)
+        {
+            if (!HasValue)
+            {
+                value = default(T);
+                return false;
+            }
+
+            value = this.value;
+            return true;
+        }
+
+        public T GetValueOrDefault(T defaultValue)
+        {
+            return HasValue ? value : defaultValue;
+        }
+
+        public override bool Equals(object other)
+        {
+            return other is Option<T> && this.Equals((Option<T>) other);
+        }
+
+        public bool Equals(Option<T> other)
+        {
+            if (!HasValue && !other.HasValue)
+            {
+                return true;
+            }
+
+            if (HasValue && other.HasValue)
+            {
+                return value.Equals(other.Value);
+            }
+
+            return false;
+        }
+
+        public static bool operator ==(Option<T> a, Option<T> b)
+        {
+            return a.Equals(b);
+        }
+
+        public static bool operator !=(Option<T> a, Option<T> b)
+        {
+            return !a.Equals(b);
+        }
+
+        public override int GetHashCode()
+        {
+            return HasValue ? value.GetHashCode() : 0;
+        }
+
+        public override string ToString()
+        {
+            return HasValue ? $"Option[{value.ToString()}]" : "Option.Empty";
+        }
+
+        public static implicit operator Option<T>(T value)
+        {
+            return new Option<T>(value);
+        }
+
+        public static implicit operator T(Option<T> option)
+        {
+            return option.Value;
+        }
+    }
+
+    public class CalledValueOnEmptyOptionException : Exception
+    {
+        public CalledValueOnEmptyOptionException()
+        {
+        }
+
+        public CalledValueOnEmptyOptionException(string message)
```

Comment:
Do we need anything other than this constructor

Reply:
We probably don't. I just thought it would be "proper style" to add the other two constructors too. See https://docs.microsoft.com/en-us/dotnet/standard/exceptions/how-to-create-user-defined-exceptions